mm2src/coins/z_coin/z_rpc.rs

```diff
@@ -1,149 +1,505 @@
+            |row| {
+                Ok(CompactBlockRow {
+                    height: BlockHeight::from_u32(row.get(0)?),
+                    data: row.get(1)?,
+                })
+            },
+        )?;
+
+        for row_result in rows {
+            let cbr = row_result?;
+            let block = CompactBlock::parse_from_bytes(&cbr.data)
+                .map_err(zcash_client_backend::data_api::error::Error::from)?;
+
+            if block.height() != cbr.height {
+                return Err(ZcashClientError::CorruptedData(format!(
+                    "Block height {} did not match row's height field value {}",
+                    block.height(),
+                    cbr.height
+                )));
+            }
+
+            with_row(block)?;
+        }
+
+        Ok(())
+    }
+
+    fn get_latest_block(&self) -> Result<i64, MmError<SqliteError>> {
+        Ok(query_single_row(
+            &self.0,
+            "SELECT height FROM compactblocks ORDER BY height DESC LIMIT 1",
+            db_common::sqlite::rusqlite::NO_PARAMS,
+            |row| row.get(0),
+        )
+        .map_err(MmError::new)?
+        .unwrap_or(0))
+    }
+
+    fn insert_block(
+        &self,
+        height: u32,
+        cb_bytes: Vec<u8>,
+    ) -> Result<usize, MmError<db_common::sqlite::rusqlite::Error>> {
+        self.0
+            .prepare("INSERT INTO compactblocks (height, data) VALUES (?, ?)")
+            .map_err(MmError::new)?
+            .execute(params![height, cb_bytes])
+            .map_err(MmError::new)
+    }
+
+    fn rewind_to_height(&self, height: u32) -> Result<usize, MmError<SqliteError>> {
+        self.0
+            .execute("DELETE from compactblocks WHERE height > ?1", [height])
+            .map_err(MmError::new)
+    }
 }
 
-#[derive(Debug, Serialize)]
-pub struct ZSendManyHtlcParams {
-    pub pubkey: H264Json,
-    pub refund_pubkey: H264Json,
-    pub secret_hash: BytesJson,
-    pub input_txid: H256Json,
-    pub input_index: usize,
-    pub input_amount: BigDecimal,
-    pub locktime: u32,
+impl BlockSource for BlockDb {
+    type Error = ZcashClientError;
+
+    fn with_blocks<F>(&self, from_height: BlockHeight, limit: Option<u32>, with_row: F) -> Result<(), Self::Error>
+    where
+        F: FnMut(CompactBlock) -> Result<(), Self::Error>,
+    {
+        self.with_blocks(from_height, limit, with_row)
+    }
 }
 
-#[derive(Debug, Deserialize)]
-pub struct ZUnspent {
-    pub txid: H256Json,
-    #[serde(rename = "outindex")]
-    pub out_index: u32,
-    pub confirmations: u32,
-    #[serde(rename = "raw_confirmations")]
-    pub raw_confirmations: Option<u32>,
-    pub spendable: bool,
-    pub address: String,
-    pub amount: MmNumber,
-    pub memo: BytesJson,
-    pub change: bool,
+pub(super) async fn init_light_client(
+    lightwalletd_url: Uri,
+    cache_db_path: impl AsRef<Path> + Send + 'static,
+    wallet_db_path: impl AsRef<Path> + Send + 'static,
```

Comment:
How about taking `cache_db_path` and `wallet_db_path` as `PathBuf`? It looks simpler and doesn't require additional compiler work.